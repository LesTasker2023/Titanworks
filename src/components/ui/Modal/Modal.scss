@import '../../../styles/design-tokens';

.modal {
  // Base overlay styles
  position: fixed;
  inset: 0;
  z-index: 50;
  background-color: rgb(0 0 0 / 0.8);
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 1rem;
  // Animation variants
  &--animation-default {
    animation: modal-fade-in 0.2s ease-out;
  }
  &--animation-fast {
    animation: modal-fade-in 0.1s ease-out;
  }
  &--animation-slow {
    animation: modal-fade-in 0.3s ease-out;
  }
  &--animation-none {
    // No animation applied
    animation: none;
  }
  // Modal content container
  &__content {
    position: relative;
    background-color: var(--background);
    border: var(--border-width-thin) solid var(--border);
    border-radius: var(--radius);
    box-shadow: var(--shadow-lg);
    width: 100%;
    max-height: 80vh;
    overflow: auto;
    // Size variants
    &--size-sm {
      max-width: 24rem;
    }
    &--size-md {
      max-width: 28rem;
    }
    &--size-lg {
      max-width: 32rem;
    }
    &--size-xl {
      max-width: 36rem;
    }
    &--size-2xl {
      max-width: 42rem;
    }
    &--size-3xl {
      max-width: 48rem;
    }
    &--size-4xl {
      max-width: 56rem;
    }
    &--size-full {
      max-width: 95vw;
    }
    // Padding variants
    &--padding-none {
      padding: 0;
    }
    &--padding-sm {
      padding: 1rem;
    }
    &--padding-md {
      padding: 1.5rem;
    }
    &--padding-lg {
      padding: 2rem;
    }
  }
  // Modal header
  &__header {
    display: flex;
    align-items: center;
    justify-content: space-between;
    padding: 1.5rem 1.5rem 0;
  }
  // Modal title
  &__title {
    font-size: 1.125rem;
    font-weight: var(--font-weight-semibold);
    line-height: 1;
    letter-spacing: -0.025em;
  }
  // Modal description
  &__description {
    font-size: 0.875rem;
    color: var(--muted-foreground);
  }
  // Modal content area
  &__content-area {
    padding: 1.5rem 1.5rem 0;
  }
  // Modal footer
  &__footer {
    display: flex;
    flex-direction: column-reverse;
    padding: 1.5rem 1.5rem 0;
    @media (min-width: 640px) {
      flex-direction: row;
      justify-content: flex-end;
      gap: 0.5rem;
    }
  }
  // Modal close button
  &__close-button {
    position: absolute;
    right: 1rem;
    top: 1rem;
    border-radius: var(--radius-sm);
    opacity: 0.7;
    background: transparent;
    border: none;
    cursor: pointer;
    transition: opacity 0.2s ease;
    &:hover {
      opacity: 1;
    }
    &:focus {
      outline: none;
      box-shadow: 0 0 0 2px var(--ring);
    }
    &:focus-visible {
      outline: var(--border-width-thick) solid var(--ring);
      outline-offset: var(--border-width-thick);
    }
    &:disabled {
      pointer-events: none;
    }
    // Close icon
    svg {
      height: 1rem;
      width: 1rem;
    }
  }
}
// Keyframe for fade-in animation
@keyframes modal-fade-in {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}
