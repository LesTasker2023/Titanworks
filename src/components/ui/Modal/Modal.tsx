import { cn } from '@/lib/utils';
import { forwardRef, useEffect } from 'react';
import { createPortal } from 'react-dom';

const getModalClassName = (animation?: string) => {
  const baseClass = 'modal';
  if (!animation || animation === 'default') return `${baseClass} ${baseClass}--animation-default`;
  return `${baseClass} ${baseClass}--animation-${animation}`;
};

const getModalContentClassName = (size?: string, padding?: string) => {
  const baseClass = 'modal__content';
  let className = baseClass;

  // Size variant
  if (size) {
    className += ` ${baseClass}--size-${size}`;
  } else {
    className += ` ${baseClass}--size-md`; // default
  }

  // Padding variant
  if (padding) {
    className += ` ${baseClass}--padding-${padding}`;
  } else {
    className += ` ${baseClass}--padding-md`; // default
  }

  return className;
};

export interface ModalProps extends React.HTMLAttributes<HTMLDivElement> {
  isOpen: boolean;
  onClose: () => void;
  children: React.ReactNode;
  animation?: 'default' | 'fast' | 'slow' | 'none';
  size?: 'sm' | 'md' | 'lg' | 'xl' | '2xl' | '3xl' | '4xl' | 'full';
  padding?: 'none' | 'sm' | 'md' | 'lg';
  closeOnOverlayClick?: boolean;
  closeOnEscape?: boolean;
  preventScroll?: boolean;
  portal?: boolean;
  portalContainer?: Element;
}

const Modal = forwardRef<HTMLDivElement, ModalProps>(
  (
    {
      className,
      animation,
      size,
      padding,
      isOpen,
      onClose,
      children,
      closeOnOverlayClick = true,
      closeOnEscape = true,
      preventScroll = true,
      portal = true,
      portalContainer,
      ...props
    },
    ref
  ) => {
    // Handle escape key
    useEffect(() => {
      if (!isOpen || !closeOnEscape) return;

      const handleEscape = (event: KeyboardEvent) => {
        if (event.key === 'Escape') {
          onClose();
        }
      };

      document.addEventListener('keydown', handleEscape);
      return () => document.removeEventListener('keydown', handleEscape);
    }, [isOpen, closeOnEscape, onClose]);

    // Handle body scroll prevention
    useEffect(() => {
      if (!isOpen || !preventScroll) return;

      const originalStyle = window.getComputedStyle(document.body).overflow;
      document.body.style.overflow = 'hidden';

      return () => {
        document.body.style.overflow = originalStyle;
      };
    }, [isOpen, preventScroll]);

    // Handle overlay click
    const handleOverlayClick = (event: React.MouseEvent<HTMLDivElement>) => {
      if (closeOnOverlayClick && event.target === event.currentTarget) {
        onClose();
      }
    };

    if (!isOpen) return null;

    const modalContent = (
      <div
        className={cn(getModalClassName(animation), className)}
        onClick={handleOverlayClick}
        role="dialog"
        aria-modal="true"
        aria-labelledby="modal-title"
        aria-describedby="modal-description"
        {...props}
        ref={ref}
      >
        <div className={getModalContentClassName(size, padding)}>{children}</div>
      </div>
    );

    if (portal) {
      const container = portalContainer || document.body;
      return createPortal(modalContent, container);
    }

    return modalContent;
  }
);

Modal.displayName = 'Modal';

// Modal sub-components for better composition
const ModalHeader = forwardRef<HTMLDivElement, React.HTMLAttributes<HTMLDivElement>>(
  ({ className, ...props }, ref) => (
    <div ref={ref} className={cn('modal__header', className)} {...props} />
  )
);
ModalHeader.displayName = 'ModalHeader';

const ModalTitle = forwardRef<HTMLHeadingElement, React.HTMLAttributes<HTMLHeadingElement>>(
  ({ className, ...props }, ref) => (
    <h2 ref={ref} id="modal-title" className={cn('modal__title', className)} {...props} />
  )
);
ModalTitle.displayName = 'ModalTitle';

const ModalDescription = forwardRef<
  HTMLParagraphElement,
  React.HTMLAttributes<HTMLParagraphElement>
>(({ className, ...props }, ref) => (
  <p ref={ref} id="modal-description" className={cn('modal__description', className)} {...props} />
));
ModalDescription.displayName = 'ModalDescription';

const ModalContent = forwardRef<HTMLDivElement, React.HTMLAttributes<HTMLDivElement>>(
  ({ className, ...props }, ref) => (
    <div ref={ref} className={cn('modal__content-area', className)} {...props} />
  )
);
ModalContent.displayName = 'ModalContent';

const ModalFooter = forwardRef<HTMLDivElement, React.HTMLAttributes<HTMLDivElement>>(
  ({ className, ...props }, ref) => (
    <div ref={ref} className={cn('modal__footer', className)} {...props} />
  )
);
ModalFooter.displayName = 'ModalFooter';

const ModalCloseButton = forwardRef<
  HTMLButtonElement,
  React.ButtonHTMLAttributes<HTMLButtonElement> & { onClose: () => void }
>(({ className, onClose, children, ...props }, ref) => (
  <button
    ref={ref}
    className={cn('modal__close-button', className)}
    onClick={onClose}
    aria-label="Close modal"
    {...props}
  >
    {children || (
      <svg
        width="15"
        height="15"
        viewBox="0 0 15 15"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m11.7816 4.03157c.0824-.07446.0824-.19618 0-.27064L11.4885 3.46826c-.0652-.05887-.1569-.05887-.2221 0L7.50002 7.2342 3.73336 3.46826c-.06516-.05887-.15698-.05887-.22214 0L3.21836 3.76093c-.08238.07446-.08238.19618 0 .27064L7.2342 8.00002 3.21836 11.9842c-.08238.0745-.08238.1962 0 .2706l.29286.2927c.06516.0589.15698.0589.22214 0L7.50002 8.7658l3.76664 3.766c.0652.0589.1569.0589.2221 0l.2929-.2927c.0824-.0744.0824-.1961 0-.2706L8.76582 8.00002 11.7816 4.03157Z"
          fill="currentColor"
          fillRule="evenodd"
          clipRule="evenodd"
        />
      </svg>
    )}
  </button>
));
ModalCloseButton.displayName = 'ModalCloseButton';

export {
  Modal,
  ModalCloseButton,
  ModalContent,
  ModalDescription,
  ModalFooter,
  ModalHeader,
  ModalTitle,
};
