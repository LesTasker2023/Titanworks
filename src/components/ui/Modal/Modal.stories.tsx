import type { Meta, StoryObj } from '@storybook/nextjs-vite';
import { useState } from 'react';
import {
  Modal,
  ModalCloseButton,
  ModalContent,
  ModalDescription,
  ModalFooter,
  ModalHeader,
  ModalTitle,
  type ModalProps,
} from './Modal';
import './Modal.scss';

const meta: Meta<typeof Modal> = {
  title: 'UI/Modal',
  component: Modal,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component:
          'A versatile modal component with accessibility features, animations, and composable sub-components.',
      },
    },
  },
  tags: ['autodocs'],
  argTypes: {
    isOpen: {
      control: { type: 'boolean' },
      description: 'Controls whether the modal is open or closed',
    },
    size: {
      control: { type: 'select' },
      options: ['sm', 'md', 'lg', 'xl', '2xl', '3xl', '4xl', 'full'],
      description: 'Size of the modal content',
    },
    padding: {
      control: { type: 'select' },
      options: ['none', 'sm', 'md', 'lg'],
      description: 'Padding inside the modal content',
    },
    animation: {
      control: { type: 'select' },
      options: ['default', 'fast', 'slow', 'none'],
      description: 'Animation speed for modal appearance',
    },
    closeOnOverlayClick: {
      control: { type: 'boolean' },
      description: 'Whether clicking the overlay closes the modal',
    },
    closeOnEscape: {
      control: { type: 'boolean' },
      description: 'Whether pressing Escape closes the modal',
    },
    preventScroll: {
      control: { type: 'boolean' },
      description: 'Whether to prevent body scroll when modal is open',
    },
  },
};

export default meta;
type Story = StoryObj<typeof meta>;

// Interactive wrapper for stories
const InteractiveModal = (args: Partial<ModalProps> & { children: React.ReactNode }) => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <div>
      <button
        onClick={() => setIsOpen(true)}
        className="px-4 py-2 bg-primary text-primary-foreground rounded-md hover:bg-primary/90"
      >
        Open Modal
      </button>
      <Modal {...args} isOpen={isOpen} onClose={() => setIsOpen(false)}>
        {args.children}
      </Modal>
    </div>
  );
};

export const Default: Story = {
  render: args => <InteractiveModal {...args} />,
  args: {
    children: (
      <>
        <ModalHeader>
          <ModalTitle>Default Modal</ModalTitle>
          <ModalCloseButton onClose={() => {}} />
        </ModalHeader>
        <ModalContent>
          <ModalDescription>
            This is a default modal with standard styling and behavior.
          </ModalDescription>
          <p className="mt-4">
            You can add any content here. The modal is fully accessible and responsive by default.
          </p>
        </ModalContent>
        <ModalFooter>
          <button className="px-4 py-2 border rounded-md hover:bg-accent">Cancel</button>
          <button className="px-4 py-2 bg-primary text-primary-foreground rounded-md hover:bg-primary/90 ml-2">
            Confirm
          </button>
        </ModalFooter>
      </>
    ),
  },
};

export const Sizes: Story = {
  render: () => {
    const [openModal, setOpenModal] = useState<string | null>(null);

    const sizes = ['sm', 'md', 'lg', 'xl', '2xl', '3xl', '4xl'] as const;

    return (
      <div className="flex flex-wrap gap-4">
        {sizes.map(size => (
          <div key={size}>
            <button
              onClick={() => setOpenModal(size)}
              className="px-4 py-2 bg-primary text-primary-foreground rounded-md hover:bg-primary/90"
            >
              {size.toUpperCase()} Modal
            </button>
            <Modal isOpen={openModal === size} onClose={() => setOpenModal(null)} size={size}>
              <ModalHeader>
                <ModalTitle>{size.toUpperCase()} Modal</ModalTitle>
                <ModalCloseButton onClose={() => setOpenModal(null)} />
              </ModalHeader>
              <ModalContent>
                <ModalDescription>
                  This is a {size} sized modal demonstrating different size variants.
                </ModalDescription>
                <p className="mt-4">
                  Modal size: <strong>{size}</strong>
                </p>
              </ModalContent>
            </Modal>
          </div>
        ))}
      </div>
    );
  },
};

export const Animations: Story = {
  render: () => {
    const [openModal, setOpenModal] = useState<string | null>(null);

    const animations = ['fast', 'default', 'slow', 'none'] as const;

    return (
      <div className="flex flex-wrap gap-4">
        {animations.map(animation => (
          <div key={animation}>
            <button
              onClick={() => setOpenModal(animation)}
              className="px-4 py-2 bg-primary text-primary-foreground rounded-md hover:bg-primary/90"
            >
              {animation} Animation
            </button>
            <Modal
              isOpen={openModal === animation}
              onClose={() => setOpenModal(null)}
              animation={animation}
            >
              <ModalHeader>
                <ModalTitle>{animation} Animation</ModalTitle>
                <ModalCloseButton onClose={() => setOpenModal(null)} />
              </ModalHeader>
              <ModalContent>
                <ModalDescription>This modal uses {animation} animation timing.</ModalDescription>
              </ModalContent>
            </Modal>
          </div>
        ))}
      </div>
    );
  },
};

export const CustomPadding: Story = {
  render: () => {
    const [openModal, setOpenModal] = useState<string | null>(null);

    const paddings = ['none', 'sm', 'md', 'lg'] as const;

    return (
      <div className="flex flex-wrap gap-4">
        {paddings.map(padding => (
          <div key={padding}>
            <button
              onClick={() => setOpenModal(padding)}
              className="px-4 py-2 bg-primary text-primary-foreground rounded-md hover:bg-primary/90"
            >
              {padding} Padding
            </button>
            <Modal
              isOpen={openModal === padding}
              onClose={() => setOpenModal(null)}
              padding={padding}
            >
              <ModalHeader>
                <ModalTitle>{padding} Padding</ModalTitle>
                <ModalCloseButton onClose={() => setOpenModal(null)} />
              </ModalHeader>
              <ModalContent>
                <ModalDescription>This modal uses {padding} padding variant.</ModalDescription>
                <div className="mt-4 p-4 bg-accent rounded">
                  Notice the different padding around this content area.
                </div>
              </ModalContent>
            </Modal>
          </div>
        ))}
      </div>
    );
  },
};

export const ConfirmationDialog: Story = {
  render: args => <InteractiveModal {...args} />,
  args: {
    size: 'sm',
    children: (
      <>
        <ModalHeader>
          <ModalTitle>Confirm Action</ModalTitle>
        </ModalHeader>
        <ModalContent>
          <ModalDescription>
            Are you sure you want to delete this item? This action cannot be undone.
          </ModalDescription>
        </ModalContent>
        <ModalFooter>
          <button className="px-4 py-2 border rounded-md hover:bg-accent">Cancel</button>
          <button className="px-4 py-2 bg-destructive text-destructive-foreground rounded-md hover:bg-destructive/90 ml-2">
            Delete
          </button>
        </ModalFooter>
      </>
    ),
  },
};

export const FormDialog: Story = {
  render: args => <InteractiveModal {...args} />,
  args: {
    size: 'lg',
    children: (
      <>
        <ModalHeader>
          <ModalTitle>Create New Project</ModalTitle>
          <ModalCloseButton onClose={() => {}} />
        </ModalHeader>
        <ModalContent>
          <ModalDescription>Fill out the form below to create a new project.</ModalDescription>
          <form className="mt-6 space-y-4">
            <div>
              <label htmlFor="project-name" className="block text-sm font-medium mb-1">
                Project Name
              </label>
              <input
                id="project-name"
                type="text"
                className="w-full p-2 border rounded-md focus:ring-2 focus:ring-primary focus:border-primary"
                placeholder="Enter project name"
              />
            </div>
            <div>
              <label htmlFor="project-description" className="block text-sm font-medium mb-1">
                Description
              </label>
              <textarea
                id="project-description"
                rows={3}
                className="w-full p-2 border rounded-md focus:ring-2 focus:ring-primary focus:border-primary"
                placeholder="Enter project description"
              />
            </div>
            <div>
              <label htmlFor="project-type" className="block text-sm font-medium mb-1">
                Project Type
              </label>
              <select
                id="project-type"
                className="w-full p-2 border rounded-md focus:ring-2 focus:ring-primary focus:border-primary"
              >
                <option>Web Application</option>
                <option>Mobile App</option>
                <option>Desktop Application</option>
                <option>API Service</option>
              </select>
            </div>
          </form>
        </ModalContent>
        <ModalFooter>
          <button className="px-4 py-2 border rounded-md hover:bg-accent">Cancel</button>
          <button className="px-4 py-2 bg-primary text-primary-foreground rounded-md hover:bg-primary/90 ml-2">
            Create Project
          </button>
        </ModalFooter>
      </>
    ),
  },
};

export const CustomBehavior: Story = {
  render: args => <InteractiveModal {...args} />,
  args: {
    closeOnOverlayClick: false,
    closeOnEscape: false,
    children: (
      <>
        <ModalHeader>
          <ModalTitle>Custom Behavior</ModalTitle>
          <ModalCloseButton onClose={() => {}} />
        </ModalHeader>
        <ModalContent>
          <ModalDescription>
            This modal cannot be closed by clicking the overlay or pressing Escape. You must use the
            close button or footer buttons.
          </ModalDescription>
          <div className="mt-4 p-4 bg-yellow-50 border border-yellow-200 rounded-md">
            <p className="text-sm text-yellow-800">
              <strong>Note:</strong> Overlay click and Escape key are disabled for this modal.
            </p>
          </div>
        </ModalContent>
        <ModalFooter>
          <button className="px-4 py-2 bg-primary text-primary-foreground rounded-md hover:bg-primary/90">
            Close Modal
          </button>
        </ModalFooter>
      </>
    ),
  },
};

export const NestedContent: Story = {
  render: args => <InteractiveModal {...args} />,
  args: {
    size: '2xl',
    children: (
      <>
        <ModalHeader>
          <ModalTitle>Rich Content Modal</ModalTitle>
          <ModalCloseButton onClose={() => {}} />
        </ModalHeader>
        <ModalContent>
          <ModalDescription>
            This modal demonstrates rich content with various UI elements.
          </ModalDescription>

          <div className="mt-6 space-y-6">
            {/* Tabs example */}
            <div>
              <h3 className="text-lg font-semibold mb-3">Project Details</h3>
              <div className="grid grid-cols-1 md:grid-cols-2 gap-4">
                <div className="space-y-2">
                  <h4 className="font-medium">General Info</h4>
                  <p className="text-sm text-muted-foreground">Project name: React Dashboard</p>
                  <p className="text-sm text-muted-foreground">Created: Jan 15, 2024</p>
                  <p className="text-sm text-muted-foreground">Status: Active</p>
                </div>
                <div className="space-y-2">
                  <h4 className="font-medium">Statistics</h4>
                  <p className="text-sm text-muted-foreground">Files: 156</p>
                  <p className="text-sm text-muted-foreground">Contributors: 8</p>
                  <p className="text-sm text-muted-foreground">Last updated: 2 hours ago</p>
                </div>
              </div>
            </div>

            {/* Progress bars */}
            <div>
              <h4 className="font-medium mb-3">Project Progress</h4>
              <div className="space-y-3">
                <div>
                  <div className="flex justify-between text-sm mb-1">
                    <span>Frontend Development</span>
                    <span>75%</span>
                  </div>
                  <div className="w-full bg-gray-200 rounded-full h-2">
                    <div className="bg-green-600 h-2 rounded-full" style={{ width: '75%' }}></div>
                  </div>
                </div>
                <div>
                  <div className="flex justify-between text-sm mb-1">
                    <span>Backend API</span>
                    <span>60%</span>
                  </div>
                  <div className="w-full bg-gray-200 rounded-full h-2">
                    <div className="bg-blue-600 h-2 rounded-full" style={{ width: '60%' }}></div>
                  </div>
                </div>
                <div>
                  <div className="flex justify-between text-sm mb-1">
                    <span>Testing</span>
                    <span>30%</span>
                  </div>
                  <div className="w-full bg-gray-200 rounded-full h-2">
                    <div className="bg-yellow-600 h-2 rounded-full" style={{ width: '30%' }}></div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </ModalContent>
        <ModalFooter>
          <button className="px-4 py-2 border rounded-md hover:bg-accent">Export Report</button>
          <button className="px-4 py-2 bg-primary text-primary-foreground rounded-md hover:bg-primary/90 ml-2">
            Save Changes
          </button>
        </ModalFooter>
      </>
    ),
  },
};
