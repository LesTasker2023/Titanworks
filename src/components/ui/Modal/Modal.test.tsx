import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { Modal, ModalContent, ModalFooter, ModalHeader } from './Modal';

describe('Modal', () => {
  // Helper component that disables portal for testing
  const BasicModal = ({ isOpen = true, portal = false, ...props }) => (
    <Modal isOpen={isOpen} onClose={() => {}} data-testid="modal" portal={portal} {...props}>
      <ModalContent>
        <ModalHeader>Modal Title</ModalHeader>
        <div>Custom content</div>
        <ModalFooter>Modal Footer</ModalFooter>
      </ModalContent>
    </Modal>
  );

  // Portal version for testing actual portal behavior
  const PortalModal = ({ isOpen = true, ...props }) => (
    <Modal isOpen={isOpen} onClose={() => {}} data-testid="modal" portal={true} {...props}>
      <ModalContent>
        <ModalHeader>Modal Title</ModalHeader>
        <div>Custom content</div>
        <ModalFooter>Modal Footer</ModalFooter>
      </ModalContent>
    </Modal>
  );

  describe('Snapshots', () => {
    it('matches default snapshot', () => {
      const { container } = render(<BasicModal />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches disabled state snapshot', () => {
      const { container } = render(<BasicModal disabled />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches different size snapshot', () => {
      const { container } = render(<BasicModal size="lg" />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches different padding snapshot', () => {
      const { container } = render(<BasicModal padding="lg" />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches different animation snapshot', () => {
      const { container } = render(<BasicModal animation="fast" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Rendering', () => {
    it('renders when open', () => {
      render(<BasicModal />);
      expect(screen.getByRole('dialog')).toBeInTheDocument();
      expect(screen.getByText('Custom content')).toBeInTheDocument();
    });

    it('does not render when closed', () => {
      render(<BasicModal isOpen={false} />);
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });

    it('renders with custom content', () => {
      render(
        <Modal isOpen onClose={() => {}} portal={false}>
          <div>Custom content</div>
        </Modal>
      );
      expect(screen.getByText('Custom content')).toBeInTheDocument();
    });

    it('renders modal header and footer', () => {
      render(<BasicModal />);
      expect(screen.getByText('Modal Title')).toBeInTheDocument();
      expect(screen.getByText('Modal Footer')).toBeInTheDocument();
    });

    it('applies custom className', () => {
      render(<BasicModal className="custom-modal" />);
      const modal = screen.getByRole('dialog');
      expect(modal).toHaveClass('custom-modal');
    });
  });

  describe('Portal Rendering', () => {
    it('renders in portal by default', () => {
      const { container } = render(<PortalModal />);
      // When using portal, container.firstChild is null
      expect(container.firstChild).toBeNull();
      // But modal is rendered in document.body
      expect(screen.getByRole('dialog')).toBeInTheDocument();
    });

    it('renders without portal when disabled', () => {
      const { container } = render(<BasicModal portal={false} />);
      expect(container.firstChild).not.toBeNull();
      expect(screen.getByRole('dialog')).toBeInTheDocument();
    });
  });

  describe('Interaction', () => {
    it('calls onClose when overlay is clicked', () => {
      const onClose = vi.fn();
      render(<BasicModal onClose={onClose} />);

      const overlay = screen.getByRole('dialog');
      fireEvent.click(overlay);
      expect(onClose).toHaveBeenCalledTimes(1);
    });

    it('does not call onClose when content is clicked', () => {
      const onClose = vi.fn();
      render(<BasicModal onClose={onClose} />);

      const content = screen.getByText('Custom content');
      fireEvent.click(content);
      expect(onClose).not.toHaveBeenCalled();
    });

    it('calls onClose when escape key is pressed', () => {
      const onClose = vi.fn();
      render(<BasicModal onClose={onClose} />);

      fireEvent.keyDown(document, { key: 'Escape' });
      expect(onClose).toHaveBeenCalledTimes(1);
    });

    it('does not call onClose on escape when closeOnEscape is false', () => {
      const onClose = vi.fn();
      render(<BasicModal onClose={onClose} closeOnEscape={false} />);

      fireEvent.keyDown(document, { key: 'Escape' });
      expect(onClose).not.toHaveBeenCalled();
    });

    it('does not call onClose on overlay click when closeOnOverlayClick is false', () => {
      const onClose = vi.fn();
      render(<BasicModal onClose={onClose} closeOnOverlayClick={false} />);

      const overlay = screen.getByRole('dialog');
      fireEvent.click(overlay);
      expect(onClose).not.toHaveBeenCalled();
    });
  });

  describe('Accessibility', () => {
    it('has correct ARIA attributes', () => {
      render(<BasicModal />);
      const modal = screen.getByRole('dialog');

      expect(modal).toHaveAttribute('aria-modal', 'true');
      expect(modal).toHaveAttribute('aria-labelledby', 'modal-title');
      expect(modal).toHaveAttribute('aria-describedby', 'modal-description');
    });

    it('has dialog role', () => {
      render(<BasicModal />);
      expect(screen.getByRole('dialog')).toBeInTheDocument();
    });
  });

  describe('Variants', () => {
    it('applies size variants correctly', () => {
      const { rerender } = render(<BasicModal size="sm" />);
      let modal = screen.getByRole('dialog');
      expect(modal.firstChild).toHaveClass('modal__content--size-sm');

      rerender(<BasicModal size="lg" />);
      modal = screen.getByRole('dialog');
      expect(modal.firstChild).toHaveClass('modal__content--size-lg');
    });

    it('applies padding variants correctly', () => {
      const { rerender } = render(<BasicModal padding="sm" />);
      let modal = screen.getByRole('dialog');
      expect(modal.firstChild).toHaveClass('modal__content--padding-sm');

      rerender(<BasicModal padding="lg" />);
      modal = screen.getByRole('dialog');
      expect(modal.firstChild).toHaveClass('modal__content--padding-lg');
    });

    it('applies animation variants correctly', () => {
      const { rerender } = render(<BasicModal animation="fast" />);
      let modal = screen.getByRole('dialog');
      expect(modal).toHaveClass('modal--animation-fast');

      rerender(<BasicModal animation="slow" />);
      modal = screen.getByRole('dialog');
      expect(modal).toHaveClass('modal--animation-slow');

      rerender(<BasicModal animation="none" />);
      modal = screen.getByRole('dialog');
      expect(modal).toHaveClass('modal--animation-none');
    });
  });

  describe('Edge Cases', () => {
    it('handles missing onClose gracefully', () => {
      // Should not throw
      expect(() => {
        render(
          <Modal isOpen onClose={undefined as never} portal={false}>
            <div>Content</div>
          </Modal>
        );
      }).not.toThrow();
    });

    it('handles empty children', () => {
      render(
        <Modal isOpen onClose={() => {}} portal={false}>
          {null}
        </Modal>
      );
      expect(screen.getByRole('dialog')).toBeInTheDocument();
    });

    it('handles multiple children', () => {
      render(
        <Modal isOpen onClose={() => {}} portal={false}>
          <div>First child</div>
          <div>Second child</div>
        </Modal>
      );
      expect(screen.getByText('First child')).toBeInTheDocument();
      expect(screen.getByText('Second child')).toBeInTheDocument();
    });
  });
});
