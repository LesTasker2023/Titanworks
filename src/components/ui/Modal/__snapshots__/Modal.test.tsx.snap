// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal > Snapshots > matches default snapshot 1`] = `
<div
  aria-describedby="modal-description"
  aria-labelledby="modal-title"
  aria-modal="true"
  class="modal modal--animation-default"
  data-testid="modal"
  role="dialog"
>
  <div
    class="modal__content modal__content--size-md modal__content--padding-md"
  >
    <div
      class="modal__content-area"
    >
      <div
        class="modal__header"
      >
        Modal Title
      </div>
      <div>
        Custom content
      </div>
      <div
        class="modal__footer"
      >
        Modal Footer
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal > Snapshots > matches different animation snapshot 1`] = `
<div
  aria-describedby="modal-description"
  aria-labelledby="modal-title"
  aria-modal="true"
  class="modal modal--animation-fast"
  data-testid="modal"
  role="dialog"
>
  <div
    class="modal__content modal__content--size-md modal__content--padding-md"
  >
    <div
      class="modal__content-area"
    >
      <div
        class="modal__header"
      >
        Modal Title
      </div>
      <div>
        Custom content
      </div>
      <div
        class="modal__footer"
      >
        Modal Footer
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal > Snapshots > matches different padding snapshot 1`] = `
<div
  aria-describedby="modal-description"
  aria-labelledby="modal-title"
  aria-modal="true"
  class="modal modal--animation-default"
  data-testid="modal"
  role="dialog"
>
  <div
    class="modal__content modal__content--size-md modal__content--padding-lg"
  >
    <div
      class="modal__content-area"
    >
      <div
        class="modal__header"
      >
        Modal Title
      </div>
      <div>
        Custom content
      </div>
      <div
        class="modal__footer"
      >
        Modal Footer
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal > Snapshots > matches different size snapshot 1`] = `
<div
  aria-describedby="modal-description"
  aria-labelledby="modal-title"
  aria-modal="true"
  class="modal modal--animation-default"
  data-testid="modal"
  role="dialog"
>
  <div
    class="modal__content modal__content--size-lg modal__content--padding-md"
  >
    <div
      class="modal__content-area"
    >
      <div
        class="modal__header"
      >
        Modal Title
      </div>
      <div>
        Custom content
      </div>
      <div
        class="modal__footer"
      >
        Modal Footer
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal > Snapshots > matches disabled state snapshot 1`] = `
<div
  aria-describedby="modal-description"
  aria-labelledby="modal-title"
  aria-modal="true"
  class="modal modal--animation-default"
  data-testid="modal"
  disabled=""
  role="dialog"
>
  <div
    class="modal__content modal__content--size-md modal__content--padding-md"
  >
    <div
      class="modal__content-area"
    >
      <div
        class="modal__header"
      >
        Modal Title
      </div>
      <div>
        Custom content
      </div>
      <div
        class="modal__footer"
      >
        Modal Footer
      </div>
    </div>
  </div>
</div>
`;
